### PR TITLE
misc: add design document for endpoints2 implementation

### PR DESCRIPTION
**Pull request by lucix-aws**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/design/endpoints.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+      "conditions": [],
+      "endpoint": {
+        "url": "https://global.api"
+      }
+    }
+  ]
+}
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
```
Comment (ianbotsf) [inline docs/design/endpoints.md:238]: **Style**: "describes how to source" → "provides"

File: docs/design/endpoints.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+public class GetResourceResolveEndpointMiddleware(
+    private val config: SmithyService.Config,
+    private val request: GetResourceRequest,
+    private val provider: EndpointProvider,
+) : ModifyRequestMiddleware {
+    override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val params = SmithyServiceEndpointParams {
+            resourceId = request.resource
+        }
```
Comment (ianbotsf) [inline docs/design/endpoints.md:282]: **Comment**: Your above definition of `SmithyServiceEndpointParams` was a data class with a public constructor and no mention of a builder whereas here it's invoked using a builder lambda. The former is preferable if we can get away with it (e.g., backward compatibility isn't a concern).

**Review by ianbotsf — CHANGES_REQUESTED** [review]

File: docs/design/endpoints.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+Client config will include a field for the resolver unique to that service, and SDK operations will use the generated
+default unless specified otherwise by the caller.
+
+### Provider library
+The endpoint rulesets as provided by Smithy require some "standard-library"-type functions that support or drive the
+evaluation of the rules.
+
+Some of these are trivial (eg. "is a value non-null") and are generated inline as part of the
+ruleset evaluation. The more complex functions (eg. URL parsing) are hosted in the runtime at
+`package aws.smithy.kotlin.runtime.http.endpoints.functions`. Runtime-hosted standard library functions are annotated as
```
Comment (aajtodd) [inline docs/design/endpoints.md:71]: Tangent tot his design, something I've wondered is whether we should pull endpoints out of `http` into it's own module (or core depending). The only hiccup here (and likely why I placed it here in the first place) is that the type relies on `Url`. Something to think about whether we want to tie endpoints to http though.

File: docs/design/endpoints.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+) {
+    /**
+     * Main codegen routine - walk the rules tree and generate the provider.
+     */
+    public fun render()
+
+    /**
+     * Resolve a standard library function referenced in a ruleset.
+     */
+    public open fun resolveFunction(func: String)
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply docs/design/endpoints.md:122]: I was wondering if this should be `Symbol` (likely)

File: docs/design/endpoints.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
+    public val resourceId: String?,
+) {
+    init {
+        require(resourceId != null) { "param resourceId is required" }
```
Comment (aajtodd) [inline docs/design/endpoints.md:249]: if it's required why is it nullable in the signature?

File: docs/design/endpoints.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
+in the case of the indices).
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
```
Comment (aajtodd) [inline docs/design/endpoints.md:245]: to Ian's point below if this is public can we guarantee ordering of parameters will be consistent over time? Otherwise this will break backwards compat one day. 

Perhaps this should follow the builder pattern for operation input/outputs?

File: docs/design/endpoints.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    public val resourceId: String?,
+) {
+    init {
+        require(resourceId != null) { "param resourceId is required" }
+    }
+}
+
+public fun interface SmithyServiceEndpointProvider {
+    open suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint
```
Comment (aajtodd) [inline docs/design/endpoints.md:254]: nit: it's an interface, `open` is implied

File: docs/design/endpoints.md [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+            // use a special endpoint for government resources
+            return Endpoint(
+                Url.parse("https://gov.api"),
+            )
+        }
+
+        // fallback to global endpoint        
+        return Endpoint(
+            Url.parse("https://global.api"),
+        )
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply docs/design/endpoints.md:270]: or something like:

```kotlin
val url = when (resourceIdPrefix) {
     "gov." -> "https://gov.api"
     else -> "https://global.api"
}
return Endpoint(Url.parse(url))
```

This is of course an implementation detail we can go over later when we actually have the full scope of rules to deal with.

File: docs/design/endpoints.md [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    }
+}
+
+public class GetResourceResolveEndpointMiddleware(
+    private val config: SmithyService.Config,
+    private val request: GetResourceRequest,
+    private val provider: EndpointProvider,
```
Comment (aajtodd) [inline docs/design/endpoints.md:277]: wouldn't this be the client specific endpoint provider `SmithyServiceEndpointProvider`? (since it takes specific parameters)

File: docs/design/endpoints.md [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        }
+
+        // fallback to global endpoint        
+        return Endpoint(
+            Url.parse("https://global.api"),
+        )
+    }
+}
+
+public class GetResourceResolveEndpointMiddleware(
```
Comment (aajtodd) [inline docs/design/endpoints.md:274]: this would be `internal`.

What package do each of these go to? I would assume this would go in the `internal` package?

File: docs/design/endpoints.md [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+```
+
+### Middleware
+Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-operation middleware.
+At runtime, endpoint resolution is comprised of two principal tasks:
+1. construct a set of provider parameters, sourcing the values as specified by the model (context params)
+2. using the sourced parameters, call the provider to resolve to an endpoint, pointing the request to it
+
+Since endpoint resolution is performed on a per-operation basis, and the sourcing of provider parameters can vary
+per-operation, middleware must now be generated per-operation (this differs from the original implementation, where the
```
Comment (aajtodd) [inline docs/design/endpoints.md:133]: Question: If an operation only uses service level context could there be a shared middleware for those operations?

This is a bit of premature optimization but we have seen how every additional class/lambda/etc adds up to overall artifact size. I would hazard a guess that most operations are not going to do anything fancy (since most don't now) and that generating a singular implementation that is shared across them would save a bunch on space/compilation/etc for larger models like DDB, EC2, S3, etc. It also would be backwards compatible because if a model ever added per/operation endpoint traits it would just generate a new internal middleware.

File: docs/design/endpoints.md [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+    init {
+        require(resourceId != null) { "param resourceId is required" }
+    }
+}
+
+public fun interface SmithyServiceEndpointProvider {
+    open suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint
+}
+
+public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider {
```
Comment (aajtodd) [inline docs/design/endpoints.md:257]: Should probable be `internal` (and live in the `internal` package) as it is now.

File: docs/design/endpoints.md [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+    public suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint {
+        val resourceIdPrefix = substring(params.resourceId, 0, 5, false)
```
Comment (aajtodd) [inline docs/design/endpoints.md:259]: Question: Are rules really going to be inlined like this? There isn't an internal data structure generated and algorithm for resolution (like there is now where we generate a partition table and have a concrete implementation in the runtime that handles resolution)? How large do we expect some rule sets to get?

File: docs/design/endpoints.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:122]: It's going to be rendering _something_ - either a symbol, or some inline logic.

File: docs/design/endpoints.md [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+### Middleware
+Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-operation middleware.
+At runtime, endpoint resolution is comprised of two principal tasks:
+1. construct a set of provider parameters, sourcing the values as specified by the model (context params)
+2. using the sourced parameters, call the provider to resolve to an endpoint, pointing the request to it
+
+Since endpoint resolution is performed on a per-operation basis, and the sourcing of provider parameters can vary
+per-operation, middleware must now be generated per-operation (this differs from the original implementation, where the
+endpoint resolver middleware was static across operations and its implementation was part of the runtime).
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:134]: There's no real way around it - the Endpoints2 spec introduces this level of variety.

File: docs/design/endpoints.md [hunk 15]
```diff
@@ -0,0 +1,292 @@
+# Endpoints
+
+* **Type**: Design
+* **Author(s)**: Luc Talatinian
+
+This document discusses the design approach to modern endpoint resolution (Endpoints v2.0) in the Kotlin SDK.
+
+## Overview
+The Smithy IDL exposes a "rules language", which allows a model author to define the programmatic steps that must be
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:9]: Sadly no, even though it's technically released in smithy now. I'm told it'll get public documentation eventually.

File: docs/design/endpoints.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:134]: Ok, seeing another comment by Aaron further down on this same thing - there _should_ be a world in which we can generate a default middleware for operations that don't have additional parameter bindings.

File: docs/design/endpoints.md (hunk 6 above)
Comment (lucix-aws) [reply docs/design/endpoints.md:245]: It _will_ generate a builder for convenience, especially since callers will be able to make these themselves (it should also simplify the codegen). I'll make sure this doc reflects that.

I'm 99% sure we would enforce struct member ordering in params like we would in any other struct, but I'll check.

File: docs/design/endpoints.md (hunk 5 above)
Comment (lucix-aws) [reply docs/design/endpoints.md:249]: Mostly to follow convention w.r.t how we do path-bound request values now (nullable, but we error on construct)

File: docs/design/endpoints.md [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+     * with those parameters, and modify the request (and handle failure to resolve if that occurs).
+     * 
+     * Client/static/general context parameters are inspected here to determine where to source values.
+     */
+    public fun render()
+    
+    /**
+     * SDK authors extend this method to handle builtins.
+     */
+    public open fun resolveBuiltin(value: String)
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:154]: It's the same issue as in `resolveFunction` where we basically have to delegate to the caller to let them handle however they need to.

These APIs should probably give a little more context about the expectations made of writer calls, to that effect.

File: docs/design/endpoints.md (hunk 13 above)
Comment (lucix-aws) [reply docs/design/endpoints.md:259]: You might have to speak more to what you're imagining but I think we're far better off codegenning directly, given that we have to evaluate arbitrary boolean expressions, assign intermediates, etc - it's just a task suited to codegen.

As far as how large, we have an s3 model with a ruleset available today. That is probably a pretty good indication of the maximum complexity we'd expect from an AWS ruleset.

File: docs/design/endpoints.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:134]: A more accurate smithy-esque term would be "bound" "bind" etc. SEP uses this verbiage. I'll update.

File: docs/design/endpoints.md [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
+      "conditions": [
+        {
+          "fn": "substring",
+          "argv": [
+            {"ref": "ResourceId"},
+            0,
+            5,
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:210]: The end index is exclusive.

File: docs/design/endpoints.md [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+  "parameters": {
+    "ResourceId": {
+      "type": "string",
+      "required": "true"
+    }
+  },
+  "rules": [
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
```
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:203]: The other type would be "error" - the fallback is just missing the type field, will fix.

File: docs/design/endpoints.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (lucix-aws) [reply docs/design/endpoints.md:270]: Yeah, I had thought about this, unfortunately we don't really have a good way to "label" the symbols - it would just have to be
```kotlin
// documentation field for 1st
object Endpoint0 : Endpoint ( ... )
// documentation field for 2nd
object Endpoint1 : Endpoint ( ...)
// etc.
```

I think exposing them publicly is IDEAL, though - especially since if people write their own resolvers, they might still want to use some of those endpoints.

I think it's worth asking if we can have the spec label them some way such that we can give them meaningful variable names.

File: docs/design/endpoints.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ianbotsf) [reply docs/design/endpoints.md:122]: Oh if it's rendering then the verb should probably reflect that.

File: docs/design/endpoints.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply docs/design/endpoints.md:270]: I disagree on exposing them unless that is something every SDK is expected to do. Anything we make public is a commitment to backwards compatibility. I'd rather retain flexibility to break the internal implementation as needed.

File: docs/design/endpoints.md (hunk 5 above)
Comment (aajtodd) [reply docs/design/endpoints.md:249]: That is modeling constraint that may or may not apply here (I can't say for sure which one without more context probably)

File: docs/design/endpoints.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (ianbotsf) [reply docs/design/endpoints.md:210]: Right and a substring from 0-4 (inclusive) would be 5 characters long:

```
"gov.foo"
 0123456
```

File: docs/design/endpoints.md [hunk 19]
```diff
@@ -63,52 +63,231 @@ Client config will include a field for the resolver unique to that service, and
...
+
+### Provider parameters
+
+Provider parameters will be generated as a data class, since backwards compatability of parameter values is guaranteed (
+field order MUST be preserved, and obsolete values are marked deprecated and not removed).
+
+The presence of required values is checked at construction, with an error thrown if any are missing.
+
+The generation includes DSL builder support for several reasons:
+1. caller convenience - SDK users are able to implement and call their own endpoint providers
```
Comment (aajtodd) [inline docs/design/endpoints.md:262]: > and call their own endpoint providers

Not sure I understand this part. Why would calling an endpoint provider be an argument for builders?

File: docs/design/endpoints.md [hunk 20]
```diff
@@ -63,52 +63,231 @@ Client config will include a field for the resolver unique to that service, and
...
-implemented to perform to walk the rules tree and generate the core logic. This base class can be extended on by SDK
+Codegen for resolvers works much like any other extendable generated construct in smithy-kotlin: a base generator is
+implemented to walk the rules tree and generate the core logic. This base class can be extended by SDK
 authors as needed.
 
 ```kotlin
 public open class EndpointProviderGenerator(
     public val writer: KotlinWriter,
-    public val ruleset: EndpointRuleset,
+    public val ruleSet: EndpointRuleset,
```
Comment (aajtodd) [inline docs/design/endpoints.md:290]: nit: simply `rules` would probably work here as well

File: docs/design/endpoints.md [hunk 21]
```diff
@@ -117,41 +296,53 @@ public open class EndpointProviderGenerator(
...
 ### Middleware
-Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-operation middleware.
-At runtime, endpoint resolution is comprised of two principal tasks:
-1. construct a set of provider parameters, sourcing the values as specified by the model (context params)
-2. using the sourced parameters, call the provider to resolve to an endpoint, pointing the request to it
 
-Since endpoint resolution is performed on a per-operation basis, and the sourcing of provider parameters can vary
-per-operation, middleware must now be generated per-operation (this differs from the original implementation, where the
-endpoint resolver middleware was static across operations and its implementation was part of the runtime).
+Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-serialization
```
Comment (aajtodd) [inline docs/design/endpoints.md:307]: if it's generating an instance of `ModifyRequestMiddleware` then it is post serialization

File: docs/design/endpoints.md [hunk 22]
```diff
@@ -271,16 +471,17 @@ public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider
     }
 }
 
-public class GetResourceResolveEndpointMiddleware(
+// package *.services.smithyservice.internal
+
+internal class GetResourceResolveEndpointMiddleware(
     private val config: SmithyService.Config,
     private val request: GetResourceRequest,
```
Comment (aajtodd) [inline docs/design/endpoints.md:478]: Something we need to be careful of here now that I'm thinking about it is that the [initialize](https://github.com/awslabs/smithy-kotlin/blob/cfa0fd3a30b4c50b75485786f043d4e2ad803f55/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt#L44) middleware phase is allowed to act on the operation input (possibly modifying it pre-serialization). We wouldn't be able to just capture the operation input at the time middleware is registered as it could change.

e.g.

```kotlin
suspend fun fooOperation(input: FooInput): FooOutput {
     val op = SdkHttpOperation<FooInput, FooOutput>.build { ... }
     
     op.install(Middleware1())
     ...
     op.install(MiddlewareN())

      // example inline middleware to demonstrate this case, same thing applies to non-inlined middleware
      op.execution.initialize.intercept { req, next ->
          val modifiedFooInput = req.subject.copy { foobar = "baz" }
          val modifiedReq = req.copy(subject = modifiedFooInput)
          next.call(modifiedReq)
      }

     // captures the original input not the one modified by the middleware that will be serialized
     op.install(FooServiceEndpointMiddleware(config, input))

}
```

Now one way around this would be to have the service endpoint middleware implement both `initialize` and `modify` phases. If it's the last `initialize` middleware to run it can capture the finalized operation input that is used during the `modify` phase.

File: docs/design/endpoints.md [hunk 23]
```diff
@@ -117,41 +296,53 @@ public open class EndpointProviderGenerator(
     public fun render()
 
     /**
-     * Resolve a standard library function referenced in a ruleset.
+     * Render a standard library function referenced in a rule set.
      */
-    public open fun resolveFunction(func: String)
+    public open fun renderFunction(func: String, args: List<FunctionArg>)
```
Comment (aajtodd) [inline docs/design/endpoints.md:301]: If these are standard library functions then I would think `resolveFunction(name: String): Symbol` would be appropriate right? The generator obviously knows the parameters already since they are a function argument so simply resolving to a symbol would be enough.

e.g.

```kotlin
// hypothetical codegen path in render()

val xyzStdlibFn = resolveFunction("xyz")
writer.write("val xyzResult = #T(param1, param2)", xyzStdlibFn)

File: docs/design/endpoints.md [hunk 24]
```diff
@@ -117,41 +296,53 @@ public open class EndpointProviderGenerator(
...
+
+    /**
+     * Implement this API to generate additional parameters to be passed to the middleware.
+     * By default, the middleware takes two parameters:
+     * 1. service client config
+     * 2. operation's request object
+     * 
+     * The implementer is responsible for code generation of the middleware installation.
+     */
+    public open fun getMiddlewareParams(): List<Symbol>
```
Comment (aajtodd) [inline docs/design/endpoints.md:339]: question: Why is this necessary? Aren't params bound to client context (config), static, or operation input?

File: docs/design/endpoints.md [hunk 25]
```diff
@@ -117,41 +296,53 @@ public open class EndpointProviderGenerator(
...
     
     /**
-     * SDK authors extend this method to handle builtins.
+     * SDK authors extend this method to handle built-ins.
+     * Called within the builder lambda for the parameters object.
      */
-    public open fun resolveBuiltin(value: String)
+    public open fun renderBuiltin(name: String)
```
Comment (aajtodd) [inline docs/design/endpoints.md:345]: Same question as above around `resolveFunction`, perhaps I'm just not seeing the bigger picture here yet...

File: docs/design/endpoints.md [hunk 26]
```diff
@@ -63,52 +63,228 @@ Client config will include a field for the resolver unique to that service, and
...
+  }
+}
+```
+
+The base middleware generator cannot render the binding of this value. The AWS SDK must implement custom logic to bind
+this value from the client config.
 
 ## Codegen
+### Package "endpoints"
+The provider parameters and interface will generate to a new top-level `endpoints` package.
```
Comment (ianbotsf) [inline docs/design/endpoints.md:252]: **Correctness**: I assume this should be "an `endpoints` package underneath each service package (e.g., `aws.sdk.kotlin.services.s3.endpoints`)", right?

File: docs/design/endpoints.md [hunk 27]
```diff
@@ -162,31 +340,35 @@ public open class EndpointResolverMiddlewareGenerator(
...
+  override fun install(op: SdkHttpOperation<GetResourceRequest, GetResourceResponse>) {
+      op.execution.initialize.intercept { req, next ->
+          // bind provider parameters and resolve endpoint, saving the value for the mutate phase
+      }
+
+      op.execution.mutate.intercept { req, next ->
+          // set the request endpoint
+      }
+  }
 }
```
Comment (ianbotsf) [inline docs/design/endpoints.md:359]: **Style**: Indentation inconsistency

File: docs/design/endpoints.md [hunk 28]
```diff
@@ -63,52 +63,228 @@ Client config will include a field for the resolver unique to that service, and
...
+@clientContextParams(
+    optInBeta: {type: "boolean", documentation: "opt in to the beta environment for this service"}
+)
+service BasicService {
+    // ...
+}
+```
+
+The client config for this service will be generated with an `optInBeta` field, the value of which will be bound for the
+corresponding endpoint parameter field at resolution.
```
Comment (ianbotsf) [inline docs/design/endpoints.md:127]: **Comment**: These new examples and expanded explanations are great. Thanks!

**Review by ianbotsf — APPROVED** [review]

**Review by aajtodd — APPROVED** [review]
